src/target/riscv/riscv.c

```diff
@@ -4886,15 +4974,19 @@ static int register_set(struct reg *reg, uint8_t *buf)
 
 	if (reg->number >= GDB_REGNO_V0 && reg->number <= GDB_REGNO_V31) {
 		if (!r->set_register_buf) {
-			LOG_ERROR("Writing register %s not supported on this RISC-V target.",
-					gdb_regno_name(reg->number));
+			LOG_TARGET_ERROR(target,
+					"Writing register %s not supported on this target.",
+					reg->name);
 			return ERROR_FAIL;
 		}
 
-		if (r->set_register_buf(target, reg->number, reg->value) != ERROR_OK)
+		if (r->set_register_buf(target, reg->number, buf) != ERROR_OK)
 			return ERROR_FAIL;
+
+		memcpy(reg->value, buf, DIV_ROUND_UP(reg->size, 8));
+		reg->valid = gdb_regno_cacheable(reg->number, /* is write? */ true);
 	} else {
-		uint64_t value = buf_get_u64(buf, 0, reg->size);
+		const riscv_reg_t value = buf_get_u64(buf, 0, reg->size);
 		if (riscv_set_register(target, reg->number, value) != ERROR_OK)
 			return ERROR_FAIL;
 	}
```

Comment:
In `riscv_set_register`, the cache is handled internally. However, `r->set_register_buf` does not contain that logic.

Could you please add it, so that these two functions are symmetric?

Reply:
I would like for accessing vector and non-vector registers to be symmetrical. But there is an issue:
`riscv_set_register` handles the cache internally, not `r->set_register`. Moreover, `r->set_register_buf` is defined in `riscv-013.c` where currently there is no cache handling at all, so I'm not sure it would be better.
One could move it to a separate function, like `riscv_set_register_buf`, but I'm not sure if it is necessary - it would be a very short function called only here.

Reply:
Ok, feel free to leave the current code as is, then. This can be refactored later.